securedrop/journalist.py

```diff
@@ -430,15 +598,85 @@ def index():
             Submission.query.filter_by(source_id=source.id,
                                        downloaded=False).all())
 
-    return render_template('index.html', unstarred=unstarred, starred=starred)
+    return starred, unstarred, source_labels, all_source_labels
+
+
+@app.route('/')
+@login_required
+def index():
+    # Long SQLAlchemy statements look best when formatted according to
+    # the Pocoo style guide, IMHO:
+    # http://www.pocoo.org/internal/styleguide/
+    sources = Source.query.filter_by(pending=False) \
+                          .order_by(Source.last_updated.desc()) \
+                          .all()
+
+    starred, unstarred, labels, all_labels = get_stars_and_labels(sources)
+
+    return render_template('index.html', unstarred=unstarred, starred=starred,
+                           source_labels=labels, all_labels=all_labels)
+
+
+@app.route('/filter_by_label/<int:label_id>', methods=('POST',))
+@login_required
+def filter_index(label_id):
+    sources = db_session.query(Source, SourceTag) \
+                        .filter_by(pending=False) \
+                        .join(SourceTag) \
+                        .filter_by(label_id=label_id) \
+                        .order_by(Source.last_updated.desc()) \
+                        .all()
+
+    sources = [x.Source for x in sources]
+    starred, unstarred, labels, all_labels = get_stars_and_labels(sources)
+
+    return render_template('index.html', unstarred=unstarred, starred=starred,
+                           source_labels=labels, all_labels=all_labels,
+                           filter_label=label_id)
 
 
 @app.route('/col/<sid>')
 @login_required
 def col(sid):
     source = get_source(sid)
+
+    # Tags on source
+    source_labels = get_source_tags(source.id)
+    selected_source_labels = [x.SourceLabelType.label_text for x in source_labels]
+    all_source_labels = get_all_defined_source_tags()
+
+    unselected_source_labels = []
+    for possible_label in all_source_labels:
+        if possible_label.label_text not in selected_source_labels:
+            unselected_source_labels.append(possible_label)
```

Comment:
If you can do this in a DB query instead of a for loop, that would be better. I bet the ORM could make this clean and expressive, but I am not sure how.